tests/test_nnsquery.c

```diff
@@ -0,0 +1,230 @@
+  int custom_size = msg_len - TCP_BASIC_SIZE;
+  if (custom_size > 0 && msg->custom) {
+    memcpy (buf, msg->custom, custom_size);
+    buf[custom_size] = '\0';
+  }
+
+  printf ("Received Topic: %s (Size: %d)\n", topic_name, msg_len);
+  printf (" - type: %d\n", msg->type);
+  printf (" - server hostname: %s\n", msg->srv_host);
+  printf (" - port: %u\n", msg->port);
+  printf (" - state: %d\n", msg->state);
+  printf (" - custom: %s\n", buf);
+}
+
+int
+tc_open_close_connection ()
+{
+  query_h handle;
+  int ret = 0;
+
+  ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);
+  if (ret != 0) {
+      printf ("Error: query_open_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(1);
+
+  ret = query_close_connection(handle);
+  if (ret != 0) {
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(1);
+
+  return 0;
+}
+
+int
+tc_send_raw_message ()
+{
+  query_h handle;
+  int ret = 0;
+  char msg[] = "test custom message";
+
+  ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);
+  if (ret != 0) {
+      printf ("Error: query_open_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  ret = query_publish_raw_data (handle, "nns_query/tcp/info", msg, strlen(msg), true);
+  if (ret != 0) {
+    printf ("Error: query_close_connection() ret: %d\n", ret);
+    return -1;
+  }
+  sleep(2);
+
+  ret = query_close_connection(handle);
+  if (ret != 0) {
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  return 0;
+}
+
+int
+tc_subscribe_raw_message ()
+{
+  query_h handle;
+  int ret = 0;
+
+  ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);
+  if (ret != 0) {
+      printf ("Error: query_open_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  ret = query_subscribe_topic (handle, "nns_query/tcp/info", raw_msg_received_cb, NULL);
+  if (ret != 0) {
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+
+  sleep (10);
+
+  ret = query_close_connection(handle);
+  if (ret != 0) {
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  return 0;
+}
+
+int
+tc_send_tcp_message ()
+{
+  query_h handle;
+  int ret = 0;
+  char msg[] = "test custom message";
+
+  ret = query_open_connection(&handle, NULL, NULL, NULL, NULL);
+  if (ret != 0) {
+      printf ("Error: query_open_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  ret = query_publish_tcp_data (handle, "nns_query/tcp/info",
+      "localhost", 12345, TCP_BUSY, msg, strlen(msg), true);
+
+  if (ret != 0) {
+    printf ("Error: query_close_connection() ret: %d\n", ret);
+    return -1;
+  }
+  sleep(2);
+
+  ret = query_close_connection(handle);
+  if (ret != 0) {
+      printf ("Error: query_close_connection() ret: %d\n", ret);
+      return -1;
+  }
+  sleep(2);
+
+  return 0;
+}
+
+int
+tc_subscribe_tcp_message ()\
```

Comment:
Unnecessary character
```suggestion
tc_subscribe_tcp_message ()
```